deploy.sh

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -p jq -i bash
+
+# shellcheck shell=bash
+
+set -euo pipefail
+
+SSHOPTS=()
+export NIX_SSHOPTS="${SSHOPTS[*]}"
+
+# For first deployment to bare server
+if [[ ${1:-} == '--prime' ]]; then
+    shift
+
+    BARE_SERVER=1
+else
+    BARE_SERVER=0
+fi
+
+get() {
+    jq -r ".$1" <<< "$MERGED"
+}
+
+check_set() {
+    if [[ "$(eval "echo \$$1")" == null ]]; then
+        echo "$2 is not set for profile $PROFILE of node $NODE"
+        exit 1
+    fi
+}
+
+deploy_profile() {
+    echo "=== Deploying profile $PROFILE of node $NODE ==="
+
+    ONLY_PROFILE="del(.path) | del(.activate)"
+    MERGED="$(jq "(. | del(.nodes) | del(.hostname) | $ONLY_PROFILE) + (.nodes.$NODE | del(.profiles) | $ONLY_PROFILE) + (.nodes.$NODE.profiles.$PROFILE | del(.hostname))" <<< "$JSON")"
+
+    HOST="$(get hostname)"
+    SSH_USER="$(get sshUser)"
+    USER="$(get user)"
+    CLOSURE="$(get path)"
+    ACTIVATE="$(get activate)"
+
+    check_set HOST hostname
+    check_set CLOSURE path
+
+    SUDO=""
+
+    if [[ "$SSH_USER" == null ]]; then
+        if [[ "$USER" == null ]]; then
+            echo "neither user nor sshUser set for profile $PROFILE of node $NODE"
+        fi
+        SSH_USER=$(whoami)
+    else
+        if [[ ! "$USER" == null ]] && [[ ! "$USER" == "$SSH_USER" ]]; then
+            SUDO="sudo -u $USER"
+        fi
+
+        if [[ "$USER" == null ]]; then
+            USER="$SSH_USER"
+        fi
+    fi
+
+
+    if [[ "$USER" == root ]]; then
+        PROFILE_PATH="/nix/var/nix/profiles/$PROFILE"
+    else
+        PROFILE_PATH="/nix/var/nix/profiles/per-user/$USER/$PROFILE"
+    fi
+
+    if [[ -n ${LOCAL_KEY:-} ]]; then
+        nix sign-paths -r -k "$LOCAL_KEY" "$CLOSURE"
+    fi
+
+    if [[ "$BARE_SERVER" == 1 ]]; then
+        echo "Checking if $HOST is up..."
+        if ! timeout 5 ssh "${SSHOPTS[@]}" "$USER@$HOST" true; then
+            echo "***** $HOST appears to be down *****"
+            read -r -p "Enter a different hostname or an IP address for $NODE ($HOST)" HOST
```

Comment:
I don't think you should prompt for a different hostname at all. Instead, ask to fix the flake or try again later.

Reply:
The idea is that e.g. for most of our repos, the machine only has an IP after initial `terraform apply`, but it also has a readable hostname. This means that IP only really makes sense for the first deployment.

Reply:
The host should already have a DNS name unless you are migrating infrastructure (which has often been the case when we split things into their `-infra` repo), but I don't see why you would have to bother with this.

You'll want to write down the IPs anyway. "write them down" in the flake and fix it later instead of prompting over and over.